src/Proto3/Wire/Reverse/Internal.hs

```diff
@@ -0,0 +1,717 @@
+
+-- | Within the monadic context established by 'allocateFields',
+-- allocates one suitably-aligned field and returns its offset.
+-- The argument is only a proxy for its type; we never evaluate it,
+-- and therefore you may pass 'undefined'.
+--
+-- WARNING: We assume that 'max' is the same as 'lcm' for any pair of
+-- alignment values, so that we can avoid using 'lcm', which does not
+-- evaluate at compile time.  Compile-time evaluation help our speed.
+allocatePrimitiveField :: Storable a => a -> State (Int, Int) Int
+allocatePrimitiveField proxy = state $ \(off0, align0) ->
+  let width1 = sizeOf proxy
+      align1 = alignment proxy
+      unaligned = off0 - width1
+      off2 = unaligned - mod unaligned align1
+      align2 = max align0 align1
+  in (off2, (off2, align2))
+
+scratchOffset, spaceOffset, stateOffset, metaDataSize, metaDataAlign :: Int
+((scratchOffset, spaceOffset, stateOffset), metaDataSize, metaDataAlign) =
+  allocateFields $
+    (,,) <$> allocatePrimitiveField (undefined :: Double)
+         <*> allocatePrimitiveField (undefined :: Int)
+         <*> allocatePrimitiveField (undefined :: Ptr ())
+               -- Note that we are allocating backward, so this
+               -- will put the pointer at the lowest address.
+
+smallChunkSize, defaultChunkSize :: Int
+smallChunkSize = BB.smallChunkSize - metaDataSize
+defaultChunkSize = BB.defaultChunkSize - metaDataSize
+
+data MetaData
+
+metaPtr :: Ptr Word8 -> Int -> Ptr MetaData
+metaPtr v = plusPtr v . negate
+
+readState :: Ptr MetaData -> IO (StablePtr (IORef BuildRState))
+readState m = castPtrToStablePtr <$> peekByteOff m stateOffset
+
+writeState :: Ptr MetaData -> StablePtr (IORef BuildRState) -> IO ()
+writeState m = pokeByteOff m stateOffset . castStablePtrToPtr
+
+readSpace :: Ptr MetaData -> IO Int
+readSpace m = peekByteOff m spaceOffset
+
+writeSpace :: Ptr MetaData -> Int -> IO ()
+writeSpace m = pokeByteOff m spaceOffset
+
+-- | The arguments are the same as the 'BuildR' arguments.
+readTotal :: Ptr Word8 -> Int -> IO Int
+readTotal v unused = do
+  -- Because we do not wish to update a record of the total
+  -- every time we write a byte, instead we record "space",
+  -- which changes rarely, and subtract "unused" from it
+  -- when we need to compute the total, which is somewhat
+  -- frequent but not as frequent as updates to "unused".
+  space <- readSpace (metaPtr v unused)
+  let !total = space - unused
+
+  -- GHC (at least v8.2.2 and v8.6.5) seems quite eager to delay the above
+  -- subtraction, even though we have indicated that the computation of
+  -- "total" is strict, and even though delaying the subtraction across
+  -- a non-inline call requires saving and restoring two registers
+  -- ("space" and "unused") instead of one ("total").  Unless we were to
+  -- completely ignore the result of the subtraction, which would be quite
+  -- unusual, an eager subtraction is faster.  Therefore we force it:
+  strictify total
```

Comment:
Would it be possible to test if `Control.Exception.evaluate` provides the same performance as `strictify`?  The reason I ask is that I think `evaluate` would be less sensitive to implementation-details of GHC.

Reply:
A good question.  I did some experiments, and sadly, it seems that `Control.Exception.evaluate` does no better than strictness annotations.  That is, the `space` figure and the `unused` figure are saved and restored, instead of just their difference, `total = space - unused`.  (After restoration the difference is computed, so unless you don't actually need the figure there is no benefit to delaying evaluation.)

The specific example I tested was the non-inline implementation of `withLengthOf`.  Before invoking the builder in the second parameter it creates a closure that includes the type information pointer of the closure, the first argument (the function from the length to the prefix builder), and the information required to compute the snapshot of the total bytes written.  With `strictify` the latter is just the total, whereas with `evaluate` it is two words that are subtracted when `withLengthOf` resumes after the builder finishes.  Thus, `strictify` shrinks the closure size from 4 words to 3, and avoids saving and reloading one word.